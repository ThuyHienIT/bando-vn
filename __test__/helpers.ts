import { randomUUID } from 'crypto';

export function generateFacility(
  type: FacilityTypeEnum,
  name: string = 'Venus Meeting Room'
): FacilityItem {
  return {
    phoneNumber: '24709620',
    id: randomUUID(),
    address: '1 JALAN KILANG,S159402',
    description:
      'Current majority suffer hair born.\nReport daughter government item keep political cut. Customer apply couple. Data happen hear must almost.',
    name,
    amenities:
      'Friend ok nation international line another within.\nUp rate when sometimes probably measure with. Product town sort. Include even alone certain.',
    type: type,
    operationHours: ['8:00', '19:00'],
    offDays: ['2022-04-23T23:18:53Z'],
    thumbnail: 'https://picsum.photos/400/200',
  };
}

export function generateBooking(facilityId: string): BookingItem {
  return {
    facilityId: facilityId,
    from: '2023-04-23T10:30:00Z',
    to: '2023-04-23T11:00:00Z',
    userEmail: 'kqthang1505@gmail.com',
    id: randomUUID(),
    createdAt: '2023-02-18T22:25:56+08:00',
    updatedAt: '2023-02-18T22:25:56+08:00',
  };
}
