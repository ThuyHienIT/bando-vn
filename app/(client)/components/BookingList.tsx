'use client';

import { Col, Empty, Row, Typography } from 'antd';
import Link from 'next/link';
import { useMemo } from 'react';
import styled from 'styled-components';

import { DoubleRightOutlined } from '@ant-design/icons';

import { BookingItem } from './BookingItem';

const ColStyle = styled(Col)``;

const ListStyle = styled(Row)`
  ${ColStyle} {
    margin-bottom: 16px;
  }
`;

const LinkRow = styled.div`
  text-align: right;
`;

interface Props {
  data: BookingItem[];
  heading: string;
  max?: number;
  viewAllLink?: string;
  className?: string;
  onCancel?(item: BookingItem): void;
  onEdit?(item: BookingItem): void;
}

export function BookingList(props: Props) {
  const listData = useMemo(() => {
    if (!props.max || !props.data || props.data.length === 0) return props.data;

    return props.data.slice(0, props.max);
  }, [props.data, props.max]);

  return (
    <div className={props.className}>
      <Typography.Title level={4}>{props.heading}</Typography.Title>
      {listData.length === 0 && (
        <Empty description={`You haven't booked any ${props.heading}`} />
      )}
      {listData.length > 0 && (
        <ListStyle gutter={16}>
          {listData.map((item) => (
            <ColStyle key={item.id} md={6}>
              <BookingItem
                data={item}
                onCancel={props.onCancel}
                onEdit={props.onEdit}
              />
            </ColStyle>
          ))}
        </ListStyle>
      )}

      {props.max && props.data.length > props.max && props.viewAllLink && (
        <LinkRow>
          <Link data-testid="view-all" href={props.viewAllLink}>
            View all <DoubleRightOutlined />
          </Link>
        </LinkRow>
      )}
    </div>
  );
}
