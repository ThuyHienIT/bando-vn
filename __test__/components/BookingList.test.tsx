import '@testing-library/jest-dom';

import { generateBooking } from '__test__/helpers';

import { BookingList } from '@components/BookingList';
import { FacilityTypeEnum } from '@enums';
import { render, screen } from '@testing-library/react';

describe('Booking List', () => {
  it('less than max', () => {
    const { container } = render(
      <BookingList
        heading="Rooms"
        max={2}
        data={[generateBooking(FacilityTypeEnum.Room)]}
        viewAllLink="/rooms"
      />
    );

    const heading = screen.getByRole('heading', { name: 'Rooms' });
    const link = screen.queryByTestId('view-all');

    expect(heading).toBeInTheDocument();
    expect(link).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it('more than max', () => {
    const { container } = render(
      <BookingList
        heading="Rooms"
        max={2}
        data={[
          generateBooking(FacilityTypeEnum.Room),
          generateBooking(FacilityTypeEnum.Room),
          generateBooking(FacilityTypeEnum.Room),
        ]}
        viewAllLink="/rooms"
      />
    );

    const link = screen.getByTestId('view-all');

    expect(link).toBeInTheDocument();
    expect(link.getAttribute('href')).toBe('/rooms');

    expect(container).toMatchSnapshot();
  });
});
