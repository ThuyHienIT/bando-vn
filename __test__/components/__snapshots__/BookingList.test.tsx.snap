// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking List less than max 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .ant-card-actions {
  margin-top: auto;
}

<div>
  <div>
    <h4
      class="ant-typography css-dev-only-do-not-override-diro6f"
    >
      Rooms
    </h4>
    <div
      class="ant-row css-dev-only-do-not-override-diro6f"
      style="margin-left: -8px; margin-right: -8px;"
    >
      <div
        class="ant-col ant-col-xs-12 ant-col-sm-8 ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c0 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-body"
          >
            <h5
              class="ant-typography css-dev-only-do-not-override-diro6f"
            />
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              From: 23/04/2023 18:30
            </span>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              To: 23/04/2023 19:00
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm ant-btn-dangerous"
                  data-testid="booking-cancel-btn"
                  type="button"
                >
                  <span>
                    Cancel
                  </span>
                </button>
              </span>
            </li>
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                  data-testid="booking-edit-btn"
                  type="button"
                >
                  <span
                    aria-label="edit"
                    class="anticon anticon-edit"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="edit"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                      />
                    </svg>
                  </span>
                  <span>
                    Edit
                  </span>
                </button>
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Booking List more than max 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .ant-card-actions {
  margin-top: auto;
}

.c1 {
  text-align: right;
}

<div>
  <div>
    <h4
      class="ant-typography css-dev-only-do-not-override-diro6f"
    >
      Rooms
    </h4>
    <div
      class="ant-row css-dev-only-do-not-override-diro6f"
      style="margin-left: -8px; margin-right: -8px;"
    >
      <div
        class="ant-col ant-col-xs-12 ant-col-sm-8 ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c0 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-body"
          >
            <h5
              class="ant-typography css-dev-only-do-not-override-diro6f"
            />
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              From: 23/04/2023 18:30
            </span>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              To: 23/04/2023 19:00
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm ant-btn-dangerous"
                  data-testid="booking-cancel-btn"
                  type="button"
                >
                  <span>
                    Cancel
                  </span>
                </button>
              </span>
            </li>
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                  data-testid="booking-edit-btn"
                  type="button"
                >
                  <span
                    aria-label="edit"
                    class="anticon anticon-edit"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="edit"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                      />
                    </svg>
                  </span>
                  <span>
                    Edit
                  </span>
                </button>
              </span>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="ant-col ant-col-xs-12 ant-col-sm-8 ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c0 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-body"
          >
            <h5
              class="ant-typography css-dev-only-do-not-override-diro6f"
            />
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              From: 23/04/2023 18:30
            </span>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              To: 23/04/2023 19:00
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm ant-btn-dangerous"
                  data-testid="booking-cancel-btn"
                  type="button"
                >
                  <span>
                    Cancel
                  </span>
                </button>
              </span>
            </li>
            <li
              style="width: 50%;"
            >
              <span>
                <button
                  class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                  data-testid="booking-edit-btn"
                  type="button"
                >
                  <span
                    aria-label="edit"
                    class="anticon anticon-edit"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="edit"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                      />
                    </svg>
                  </span>
                  <span>
                    Edit
                  </span>
                </button>
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="c1"
    >
      <a
        data-testid="view-all"
        href="/rooms"
      >
        View all 
        <span
          aria-label="double-right"
          class="anticon anticon-double-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="double-right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 00188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 00492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
            />
          </svg>
        </span>
      </a>
    </div>
  </div>
</div>
`;
